src/open_r1/rewards.py

```diff
@@ -450,43 +472,35 @@ def evaluate_code(code, test_cases):
 
     evaluate_code(code_snippet, test_cases)
     """
+
     code_snippets = [extract_code(completion[-1]["content"]) for completion in completions]
     verification_info = kwargs["verification_info"]
+    
+    # Choose the appropriate template based on provider type
+    template = evaluation_script_template
+    
     scripts = [
-        evaluation_script_template.format(code=json.dumps(code), test_cases=json.dumps(json.dumps(info["test_cases"])))
+        template.format(
+            code=json.dumps(code), 
+            test_cases=json.dumps(json.dumps(info["test_cases"]))
+        )
         for code, info in zip(code_snippets, verification_info)
     ]
 
     language = verification_info[0]["language"]
-    if not all(v["language"] == language for v in verification_info):
-        raise ValueError("All verification_info must have the same language", verification_info)
+    
+    if enforce_same_language:
+        all_same_language = all(v["language"] == language for v in verification_info)
+        if not all_same_language:
+            raise ValueError("All verification_info must have the same language", verification_info)
 
-    if e2b_router_url is not None:
-        routed_sandbox = RoutedSandbox(router_url=e2b_router_url)
-
-        executions = routed_sandbox.run_code(
-            scripts=scripts,
-            language=language,
-            timeout=30,
-            request_timeout=28,
-        )
-
-        rewards = []
-        for execution in executions:
-            try:
-                reward = float(execution.text)
-                rewards.append(reward)
-            except Exception:
-                rewards.append(None)
-        return rewards
-
-    try:
-        rewards = run_async_from_sync(scripts, language, num_parallel)
-    except Exception as e:
-        print(f"Error from E2B executor: {e}")
-        rewards = [0.0] * len(completions)
-
-    return rewards
+    execution_provider = get_provider(
+        provider_type=provider_type,
+        num_parallel=num_parallel,
+        e2b_router_url=e2b_router_url,
```

Comment:
We should propagate `kwargs` here to use Morph right?
```suggestion
        **kwargs,
```